### DISCUSSION TITLE
tutorials/scripting/gdscript/static_typing

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#98] [User-contributed notes]
# tutorials/scripting/gdscript/static_typing

In this guide, you will learn: how to use static typing in GDScript;, that static types can help you avoid bugs;, that static typing improves your experience with the editor.. Where and how you use...

https://docs.godotengine.org/en/4.3/tutorials/scripting/gdscript/static_typing.html

**Comment by edumaxsantos** [comment]
any plans on supporting typed dictionaries?

**Reply by Calinou** [reply]
Yes, see https://github.com/godotengine/godot-proposals/issues/56 :slightly_smiling_face:

**Reply by Lazy-Rabbit-2001** [reply]
Typed dictionaries are implemented in 4.4, so it's time to update this page, imo

**Reply by anzbert** [reply]
hope 4.4 will be stable soon :) its a huge improvement

**Comment by KleidonBD** [comment]
Why does the "Specify the element type of an Array" section say that "var arrays: Array[Array]" is disallowed? You can statically type the elements of an Array to also be Array, you just can't have the nested Array be statically typed.

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/10095

**Comment by Werninator** [comment]
Typed Dictionaries are available as of 4.4beta3, go ham! They're awesome :D

**Comment by Stan-Stani** [comment]
Typed Dictionary Syntax: `var view_states: Dictionary[int, float] = {}`

**Reply by Calinou** [reply]
This is already mentioned in the `latest` version of the documentation: https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/static_typing.html#specify-the-element-type-of-a-dictionary

It's not cherry-picked to `stable` yet as per https://github.com/godotengine/godot-docs/issues/10981.

**Comment by ghost** [comment]
Beware typed constructors. If returning from a function an empty value, for example, the most performant option to construct the empty dictionary is using brackets, not the constructor. In a 10kk loop, the differences are as follows:
```GDScript
# 2230 msec
func get_dict() -> Dictionary[String, int]:
    return Dictionary(Dictionary(), TYPE_STRING, StringName(), null, TYPE_INT, StringName(), null)
```
```GDScript
# 1916 msec
func get_dict() -> Dictionary[String, int]:
    return {}
```
I assume it's because in the typed constructor we first create an untyped Dictionary, as well as having to create two StringNames, before actually calling the constructor itself.

When it comes to array subtypes, the situation is the exact opposite:
```GDScript
# 459 msec
func get_arr() -> PackedStringArray:
    return PackedStringArray()
```
```GDScript
# 1014 msec
func get_arr() -> PackedStringArray:
    return []
```
This because using brackets first creates an Array which then gets converted to a PackedStringArray via the constructor. Basically, it's the equivalent of writing `PackedStringArray([])`.

For other cases such as `String()` and `""` or `StringName()` and `&""` I haven't found a significant difference.

**Comment by salandered** [comment]
> A safer alternative is to use the is keyword:
> ```if not (body is PlayerController):```
> ...
> You can also simplify the code by using the is not operator:
> ```if body is not PlayerController:```

Just in case, you can also use ```if not body is PlayerController``` ('is' has higher precedence)
